Mathlib/MeasureTheory/Integral/Indicator.lean

```diff
@@ -0,0 +1,226 @@
+   (for the ones using `AEStronglyMeasurable`)
+
+Those in section TendstoMeasureOfTendstoIndicator (except to the extent they rely on measurability)
+only have prerequisites
+ * `Mathlib.MeasureTheory.Integral.Lebesgue`
+ -/
+
+open MeasureTheory Set Filter Topology ENNReal NNReal BigOperators
+
+section IndicatorConstMeasurable
+
+-- I didn't find the following lemma, the closest variants were around `indicator_const_preimage`.
+--#check indicator_const_preimage
+
+/-- A characterization of the measurability of the indicator function which takes a constant
+value `b` on a set `A` and `0` elsewhere. (This version requires the measurability of the singleton
+`{0}` as an explicit input, see `measurable_indicator_const_iff` for a version with typeclass
+inference.) -/
+lemma measurable_indicator_const_iff' [MeasurableSpace α] (A : Set α) [Zero β] [MeasurableSpace β]
+  (h0 : MeasurableSet ({0} : Set β)) (b : β) :
+    Measurable (A.indicator (fun _ ↦ b)) ↔ (b = 0 ∨ MeasurableSet A) := by
+  constructor <;> intro h
+  · by_cases hb : b = 0 <;> simp only [hb, true_or, false_or]
+    convert h h0.compl
+    ext a
+    simp [hb]
+  · by_cases hb : b = 0
+    · simp only [hb, indicator_zero, measurable_const]
+    · have A_mble : MeasurableSet A := by simpa only [hb, false_or] using h
+      intro B _
+      rcases indicator_const_preimage A B b with ⟨hB⟩ | ⟨hB | ⟨hB | hB⟩⟩
+      · simp only [hB, MeasurableSet.univ]
+      · simp only [hB, A_mble]
+      · simp only [hB, MeasurableSet.compl_iff, A_mble]
+      · simp only [mem_singleton_iff] at hB
+        simp only [hB, MeasurableSet.empty]
+
+--#find_home measurable_indicator_const_iff'
+-- Gives: `Mathlib.MeasureTheory.Integral.Indicator`, i.e., this file itself...
+-- But why? Could be in `Mathlib.MeasureTheory.Constructions.BorelSpace.Metrizable`!
+
+/-- A characterization of the measurability of the indicator function which takes a constant
+value `b` on a set `A` and `0` elsewhere. -/
+lemma measurable_indicator_const_iff [MeasurableSpace α] (A : Set α) [Zero β] [MeasurableSpace β]
+  [MeasurableSingletonClass β] (b : β) :
+    Measurable (A.indicator (fun _ ↦ b)) ↔ (b = 0 ∨ MeasurableSet A) :=
+  measurable_indicator_const_iff' A (MeasurableSet.singleton 0) b
+
+/-- A characterization of the a.e.-measurability of the indicator function which takes a constant
+value `b` on a set `A` and `0` elsewhere. (This version requires the measurability of the singleton
+`{0}` as an explicit input, see `measurable_indicator_const_iff` for a version with typeclass
+inference.) -/
+lemma aeMeasurable_indicator_const_iff' [MeasurableSpace α] (A : Set α) [DecidableEq β]
+  [Zero β] [MeasurableSpace β] [TopologicalSpace β] [TopologicalSpace.PseudoMetrizableSpace β]
+  [BorelSpace β]
+  [TopologicalSpace.SecondCountableTopology β] [OpensMeasurableSpace β] (μ : Measure α)
+  (h0 : MeasurableSet ({0} : Set β)) (b : β) :
+    AEStronglyMeasurable (A.indicator (fun _ ↦ b)) μ ↔ (b = 0 ∨ NullMeasurableSet A μ) := by
```

Comment:
Golfing this proof would be very welcome! (Also I am not sure if some variant of the statement would be more helpful instead.)

Reply:
As I suggested above, I think a first step would be to split out the case where `b` is nonzero. Also, I'd say this is fundamentally a statement about `AEMeasurable`, not `AEStronglyMeasurable`, and it would probably save you some troubles.

Other than that, it's generally best to avoid unfolding the definitions of `AEMeasurable` and `NullMeasurableSet` when you can, and I think [AEMeasurable.nullMeasurable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.nullMeasurable) allows you to avoid that when getting measurability of the set from measurability of the indicator. For the other direction, I suggest modifying [AEMeasurable.indicator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.indicator) to work with a null measurable set, using [indicator_eventuallyEq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/IndicatorFunction.html#indicator_eventuallyEq)

Reply:
I changed this to about` AEMeasurable`.

Generalizing [AEMeasurable.indicator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.indicator) to null-measurable instead of measurable sets seems to require generalizing the chain <- [aemeasurable_indicator_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#aemeasurable_indicator_iff) <- [indicator_ae_eq_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#indicator_ae_eq_restrict) <- [piecewise_ae_eq_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#piecewise_ae_eq_restrict) <- [ae_restrict_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.ae_restrict_eq) <- [restrict_apply_eq_zero'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_apply_eq_zero') <- [restrict_apply'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_apply') (<- [restrict_toOuterMeasure_eq_toOuterMeasure_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_toOuterMeasure_eq_toOuterMeasure_restrict) <- ...). That might be a worthwhile refactor, but in #6253 I would only need the measurable sets case, so I'd prefer to either keep these more ad hoc proofs or remove the null-measurable and a.e.-measurable versions altogether.

Reply:
Thank you very much, this is great! I also realized that contrary to what I said above, I probably do want the a.e.-measurable/null-measurable versions after all, so I will be extremely happy if they are not removed from the PR.

I don't know if I marked the "depends on #6445" correctly, since this PR didn't become automatically marked "blocked by another PR". Do I need to add that label by hand when I add a "depends on" not at the PR creation but later?

Reply:
If [this message](https://github.com/leanprover-community/mathlib4/pull/6225#issuecomment-1670885097) appears then it's correctly ported. I think the bot is supposed to add the label by itself, but sometimes it takes some time / doesn't work.